// @aws-cdk/aws-kinesisfirehose/lib/lambda-function-processor.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaFunctionProcessor = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * (experimental) Use an AWS Lambda function to transform records.
 *
 * @experimental
 */
class LambdaFunctionProcessor {
    /**
     * @experimental
     */
    constructor(lambdaFunction, props = {}) {
        this.lambdaFunction = lambdaFunction;
        jsiiDeprecationWarnings._aws_cdk_aws_kinesisfirehose_DataProcessorProps(props);
        this.props = props;
    }
    /**
     * (experimental) Binds this processor to a destination of a delivery stream.
     *
     * Implementers should use this method to grant processor invocation permissions to the provided stream and return the
     * necessary configuration to register as a processor.
     *
     * @experimental
     */
    bind(_scope, options) {
        jsiiDeprecationWarnings._aws_cdk_aws_kinesisfirehose_DataProcessorBindOptions(options);
        this.lambdaFunction.grantInvoke(options.role);
        return {
            processorType: 'Lambda',
            processorIdentifier: {
                parameterName: 'LambdaArn',
                parameterValue: this.lambdaFunction.functionArn,
            },
        };
    }
}
exports.LambdaFunctionProcessor = LambdaFunctionProcessor;
_a = JSII_RTTI_SYMBOL_1;
LambdaFunctionProcessor[_a] = { fqn: "@aws-cdk/aws-kinesisfirehose.LambdaFunctionProcessor", version: "1.137.0" };
