// @aws-cdk/aws-ecs/lib/log-drivers/log-drivers.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogDrivers = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_log_driver_1 = require("./aws-log-driver");
const firelens_log_driver_1 = require("./firelens-log-driver");
const fluentd_log_driver_1 = require("./fluentd-log-driver");
const gelf_log_driver_1 = require("./gelf-log-driver");
const journald_log_driver_1 = require("./journald-log-driver");
const json_file_log_driver_1 = require("./json-file-log-driver");
const splunk_log_driver_1 = require("./splunk-log-driver");
const syslog_log_driver_1 = require("./syslog-log-driver");
/**
 * The base class for log drivers.
 *
 * @stability stable
 */
class LogDrivers {
    /**
     * Creates a log driver configuration that sends log information to CloudWatch Logs.
     *
     * @stability stable
     */
    static awsLogs(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_AwsLogDriverProps(props);
        return new aws_log_driver_1.AwsLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to fluentd Logs.
     *
     * @stability stable
     */
    static fluentd(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_FluentdLogDriverProps(props);
        return new fluentd_log_driver_1.FluentdLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to gelf Logs.
     *
     * @stability stable
     */
    static gelf(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_GelfLogDriverProps(props);
        return new gelf_log_driver_1.GelfLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to journald Logs.
     *
     * @stability stable
     */
    static journald(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_JournaldLogDriverProps(props);
        return new journald_log_driver_1.JournaldLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to json-file Logs.
     *
     * @stability stable
     */
    static jsonFile(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_JsonFileLogDriverProps(props);
        return new json_file_log_driver_1.JsonFileLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to splunk Logs.
     *
     * @stability stable
     */
    static splunk(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_SplunkLogDriverProps(props);
        return new splunk_log_driver_1.SplunkLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to syslog Logs.
     *
     * @stability stable
     */
    static syslog(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_SyslogLogDriverProps(props);
        return new syslog_log_driver_1.SyslogLogDriver(props);
    }
    /**
     * Creates a log driver configuration that sends log information to firelens log router.
     *
     * For detail configurations, please refer to Amazon ECS FireLens Examples:
     * https://github.com/aws-samples/amazon-ecs-firelens-examples
     *
     * @stability stable
     */
    static firelens(props) {
        jsiiDeprecationWarnings._aws_cdk_aws_ecs_FireLensLogDriverProps(props);
        return new firelens_log_driver_1.FireLensLogDriver(props);
    }
}
exports.LogDrivers = LogDrivers;
_a = JSII_RTTI_SYMBOL_1;
LogDrivers[_a] = { fqn: "@aws-cdk/aws-ecs.LogDrivers", version: "1.137.0" };
