// @aws-cdk/aws-codebuild/lib/events.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhaseChangeEvent = exports.StateChangeEvent = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const events = require("@aws-cdk/aws-events");
/**
 * Event fields for the CodeBuild "state change" event.
 *
 * @see https://docs.aws.amazon.com/codebuild/latest/userguide/sample-build-notifications.html#sample-build-notifications-ref
 * @stability stable
 */
class StateChangeEvent {
    constructor() {
    }
    /**
     * The triggering build's status.
     *
     * @stability stable
     */
    static get buildStatus() {
        return events.EventField.fromPath('$.detail.build-status');
    }
    /**
     * The triggering build's project name.
     *
     * @stability stable
     */
    static get projectName() {
        return events.EventField.fromPath('$.detail.project-name');
    }
    /**
     * Return the build id.
     *
     * @stability stable
     */
    static get buildId() {
        return events.EventField.fromPath('$.detail.build-id');
    }
    /**
     * @stability stable
     */
    static get currentPhase() {
        return events.EventField.fromPath('$.detail.current-phase');
    }
}
exports.StateChangeEvent = StateChangeEvent;
_a = JSII_RTTI_SYMBOL_1;
StateChangeEvent[_a] = { fqn: "@aws-cdk/aws-codebuild.StateChangeEvent", version: "1.137.0" };
/**
 * Event fields for the CodeBuild "phase change" event.
 *
 * @see https://docs.aws.amazon.com/codebuild/latest/userguide/sample-build-notifications.html#sample-build-notifications-ref
 * @stability stable
 */
class PhaseChangeEvent {
    constructor() {
    }
    /**
     * The triggering build's project name.
     *
     * @stability stable
     */
    static get projectName() {
        return events.EventField.fromPath('$.detail.project-name');
    }
    /**
     * The triggering build's id.
     *
     * @stability stable
     */
    static get buildId() {
        return events.EventField.fromPath('$.detail.build-id');
    }
    /**
     * The phase that was just completed.
     *
     * @stability stable
     */
    static get completedPhase() {
        return events.EventField.fromPath('$.detail.completed-phase');
    }
    /**
     * The status of the completed phase.
     *
     * @stability stable
     */
    static get completedPhaseStatus() {
        return events.EventField.fromPath('$.detail.completed-phase-status');
    }
    /**
     * The duration of the completed phase.
     *
     * @stability stable
     */
    static get completedPhaseDurationSeconds() {
        return events.EventField.fromPath('$.detail.completed-phase-duration-seconds');
    }
    /**
     * Whether the build is complete.
     *
     * @stability stable
     */
    static get buildComplete() {
        return events.EventField.fromPath('$.detail.build-complete');
    }
}
exports.PhaseChangeEvent = PhaseChangeEvent;
_b = JSII_RTTI_SYMBOL_1;
PhaseChangeEvent[_b] = { fqn: "@aws-cdk/aws-codebuild.PhaseChangeEvent", version: "1.137.0" };
